XpressAI/xircuits
# 📖 Quick Start Pytorch Components & Workflows

# Description

With the upcoming 1,5 release, we're looking forward to improve the new user onboarding. This PR introduces a new set of pytorch components based on the official [pytorch tutorials](https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html). Most of the components are designed to emulate commonly used modules such as torch utils `Dataset` and `DataLoader`, while the rest such as `TorchModel` are easily modifiable. 

List of new components:

1. LoadTorchVisionDataset
2. TorchDataLoader
3. TorchModel
4. TrainTorchModel
5. TestTorchModel
6. SaveTorchModelState
7. LoadTorchModelState
8. TorchModelPredict
9. Image2TorchTensor
10. TorchModelPredictFromTensor

I have also created 2 workflows:

- **TorchTrainModel.xircuits**: Trains a fashion mnist model, and saves the trained torch model state.
- **TrainPredictFromModel.xircuits**: Loads a saved torch model state and performs inference.

![image](https://user-images.githubusercontent.com/68586800/180361685-7ce30712-edf2-445d-86bc-d70b4d6d97a7.png)

All components are complete with their descriptions.

Due to it's specific use case, the previous pytorch components and workflows are migrated to it's own Xircuits project template - https://github.com/XpressAI/template-semantic_segmentation

## References

https://github.com/XpressAI/xircuits/pull/68

## Pull Request Type


## Type of Change


# Tests
1. Setup
    1. Make venv
    ```python -m venv venv && source venv/Scripts/activate ```
    2. After extracting the zip file
    ```pip install xircuits*.whl```
    3. To fetch the new torch components:
    ```xircuits-components --branch fahreza/new-torch```
    4. pip install the new torch components
    ```pip install -r xai_components/xai_pytorch/requirements.txt```
    5. Launch Xircuits:
    ```xircuits```


2. Workflow Tests
   1. Run `TorchTrainModel.xircuits`. Did it successfully run?
   3. Run `TrainPredictFromModel.xircuits`. Did it successfully run?

2. Grammar Check
Check that the new components have reasonable text descriptions.


## Tested on?


# Notes

The prediction isn't great since the model accuracy is only 60%. Feel free to add your own if it gives the correct output.
